File: src/shell/builtin/exit.zig
state: enum {
    idle,
    waiting_io,
    err,
    done,
} = .idle,

pub fn start(this: *Exit) Yield {
    const args = this.bltn().argsSlice();
    switch (args.len) {
        0 => {
            return this.bltn().done(0);
        },
        1 => {
            const first_arg = args[0][0..std.mem.len(args[0]) :0];
            const exit_code: ExitCode = std.fmt.parseInt(u8, first_arg, 10) catch |err| switch (err) {
                error.Overflow => @intCast((std.fmt.parseInt(usize, first_arg, 10) catch return this.fail("exit: numeric argument required\n")) % 256),
                error.InvalidCharacter => return this.fail("exit: numeric argument required\n"),
            };
            return this.bltn().done(exit_code);
        },
        else => {
            return this.fail("exit: too many arguments\n");
        },
    }
}

fn fail(this: *Exit, msg: []const u8) Yield {
    if (this.bltn().stderr.needsIO()) |safeguard| {
        this.state = .waiting_io;
        return this.bltn().stderr.enqueue(this, msg, safeguard);
    }
    _ = this.bltn().writeNoIO(.stderr, msg);
    return this.bltn().done(1);
}

pub fn next(this: *Exit) Yield {
    switch (this.state) {
        .idle => shell.unreachableState("Exit.next", "idle"),
        .waiting_io => {
            return .suspended;
        },
        .err => {
            return this.bltn().done(1);
        },
        .done => {
            return this.bltn().done(1);
        },
    }
}

pub fn onIOWriterChunk(this: *Exit, _: usize, maybe_e: ?JSC.SystemError) Yield {
    if (comptime bun.Environment.allow_assert) {
        assert(this.state == .waiting_io);
    }
    if (maybe_e) |e| {
        defer e.deref();
        this.state = .err;
        return this.next();
    }
    this.state = .done;
    return this.next();
}

pub fn deinit(this: *Exit) void {
    _ = this;
}

pub inline fn bltn(this: *Exit) *Builtin {
    const impl: *Builtin.Impl = @alignCast(@fieldParentPtr("exit", this));
    return @fieldParentPtr("impl", impl);
}

// --
const bun = @import("bun");
const shell = bun.shell;
const Yield = shell.Yield;
const interpreter = @import("../interpreter.zig");
const Interpreter = interpreter.Interpreter;
const Builtin = Interpreter.Builtin;
const ExitCode = shell.ExitCode;
const Exit = @This();
const JSC = bun.JSC;
const std = @import("std");

const assert = bun.assert;

File: src/runtime/shell/builtin/exit.rs
use crate::shell::builtin::{Builtin, BuiltinState};
use crate::shell::interpreter::{Interpreter, NodeId};
use crate::shell::yield_::Yield;

#[derive(Default)]
pub struct Exit {
    state: State,
}

#[derive(Default)]
enum State {
    #[default]
    Idle,
    WaitingIo,
    Done,
}

impl Exit {
    pub(crate) fn start(interp: &Interpreter, cmd: NodeId) -> Yield {
        let code: Result<crate::shell::ExitCode, &'static [u8]> = {
            let bltn = Builtin::of(interp, cmd);
            match bltn.args_slice().len() {
                0 => Ok(0),
                1 => parse_exit_code(bltn.arg_bytes(0))
                    .ok_or(b"exit: numeric argument required\n".as_slice()),
                _ => Err(b"exit: too many arguments\n".as_slice()),
            }
        };
        let code = match code {
            Ok(c) => c,
            Err(msg) => return Self::fail(interp, cmd, msg),
        };
        // Intentional divergence from bash: this completes only the current
        // Cmd rather than unwinding the whole script.
        Builtin::done(interp, cmd, code)
    }

    fn fail(interp: &Interpreter, cmd: NodeId, msg: &[u8]) -> Yield {
        Self::state_mut(interp, cmd).state = State::WaitingIo;
        Builtin::write_failing_error(interp, cmd, msg, 1)
    }

    pub(crate) fn on_io_writer_chunk(
        interp: &Interpreter,
        cmd: NodeId,
        _: usize,
        _err: Option<bun_sys::SystemError>,
    ) -> Yield {
        Self::state_mut(interp, cmd).state = State::Done;
        Builtin::done(interp, cmd, 1)
    }
}

fn parse_exit_code(s: &[u8]) -> Option<crate::shell::ExitCode> {
    // %256 is bash semantics — keep wrapper fn.
    bun_core::fmt::parse_decimal::<u64>(s).map(|n| (n % 256) as crate::shell::ExitCode)
}
